crates/ruma-common/src/canonical_json.rs

```diff
@@ -0,0 +1,296 @@
+//! Canonical JSON types and related functions.
+
+use std::{fmt, mem};
+
+use serde::Serialize;
+use serde_json::Value as JsonValue;
+
+mod value;
+
+use crate::RoomVersionId;
+
+pub use self::value::{CanonicalJsonObject, CanonicalJsonValue};
+
+/// The set of possible errors when serializing to canonical JSON.
+#[cfg(feature = "canonical-json")]
+#[derive(Debug)]
+#[allow(clippy::exhaustive_enums)]
+pub enum CanonicalJsonError {
+    /// The numeric value failed conversion to js_int::Int.
+    IntConvert,
+
+    /// An error occurred while serializing/deserializing.
+    SerDe(serde_json::Error),
+}
+
+impl fmt::Display for CanonicalJsonError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            CanonicalJsonError::IntConvert => {
+                f.write_str("number found is not a valid `js_int::Int`")
+            }
+            CanonicalJsonError::SerDe(err) => write!(f, "serde Error: {err}"),
+        }
+    }
+}
+
+impl std::error::Error for CanonicalJsonError {}
+
+/// Fallible conversion from a `serde_json::Map` to a `CanonicalJsonObject`.
+pub fn try_from_json_map(
+    json: serde_json::Map<String, JsonValue>,
+) -> Result<CanonicalJsonObject, CanonicalJsonError> {
+    json.into_iter().map(|(k, v)| Ok((k, v.try_into()?))).collect()
+}
+
+/// Fallible conversion from any value that impl's `Serialize` to a `CanonicalJsonValue`.
+pub fn to_canonical_value<T: Serialize>(
+    value: T,
+) -> Result<value::CanonicalJsonValue, CanonicalJsonError> {
+    serde_json::to_value(value).map_err(CanonicalJsonError::SerDe)?.try_into()
+}
+
+/// Redacts an event using the rules specified in the Matrix client-server specification.
+///
+/// This is part of the process of signing an event.
+///
+/// Redaction is also suggested when a verifying an event with `verify_event` returns
+/// `Verified::Signatures`. See the documentation for `Verified` for details.
+///
+/// Returns a new JSON object with all applicable fields redacted.
+///
+/// # Parameters
+///
+/// * object: A JSON object to redact.
+///
+/// # Errors
+///
+/// Returns an error if:
+///
+/// * `object` contains a field called `content` that is not a JSON object.
+/// * `object` contains a field called `hashes` that is not a JSON object.
+/// * `object` contains a field called `signatures` that is not a JSON object.
+/// * `object` is missing the `type` field or the field is not a JSON string.
+pub fn redact(
+    object: &CanonicalJsonObject,
+    version: &RoomVersionId,
+) -> Result<CanonicalJsonObject, ()> {
+    let mut val = object.clone();
+    redact_in_place(&mut val, version)?;
+    Ok(val)
+}
+
+/// Redacts an event using the rules specified in the Matrix client-server specification.
+///
+/// Functionally equivalent to `redact`, only;
+/// * upon error, the event is not touched.
+/// * this'll redact the event in-place.
+pub fn redact_in_place(event: &mut CanonicalJsonObject, version: &RoomVersionId) -> Result<(), ()> {
```

Comment:
The error type here is the main reason this is still a work-in-progress. I'm really not sure what it should be: The error type in ruma-signatures is very detailed, but I'm pretty skeptical about its usefulness. Maybe we should just have an opaque error type that you can only print? In all cases, it means that the event is malformed in some way.

Reply:
I think that the ruma-signatures error is just a crate error, and that only a few of those are dedicated to redaction.

We could just incorperate them into the ruma-common error type, or make a specific error type to canonical json / redaction stuff.